File: nudge-next/src/lib/auth/rbac.ts
/**
 * Role-based access control for workspace members.
 *
 * 1:1 TypeScript port of backend/auth/rbac.py. The five roles use the Clerk org
 * role keys (`org:owner` … `org:viewer`) as the canonical internal identifiers,
 * and the eleven permissions use the Clerk custom-permission keys.
 *
 *   org:owner    full control incl. billing, delete workspace, manage admins
 *   org:admin    manage team, campaigns, templates, settings
 *   org:manager  view all leads, assign leads, see reports
 *   org:agent    handle only leads/conversations assigned to them
 *   org:viewer   read-only
 *
 * Pure module — no Appwrite, no Clerk SDK, no `server-only`. Safe to import
 * anywhere (UI RoleGate, Server Actions, Appwrite Functions).
 */

import type { Role, Permission } from "@/types/roles";

export type { Role, Permission };

// Privilege rank — used to stop a member from managing someone above them.
export const ROLE_RANK: Record<Role, number> = {
  "org:owner": 4,
  "org:admin": 3,
  "org:manager": 2,
  "org:agent": 1,
  "org:viewer": 0,
};

export const ROLES = Object.keys(ROLE_RANK) as Role[];

// Permission catalogue (granular capabilities used by endpoints / actions).
export const P_BILLING: Permission = "org:billing:manage";
export const P_WORKSPACE_EDIT: Permission = "org:workspace:edit";
export const P_WORKSPACE_DEL: Permission = "org:workspace:delete";
export const P_TEAM_MANAGE: Permission = "org:team:manage"; // invite/remove members, change roles, sales teams
export const P_CAMPAIGNS: Permission = "org:campaigns:manage";
export const P_TEMPLATES: Permission = "org:templates:manage";
export const P_SETTINGS: Permission = "org:settings:manage"; // vertical, WhatsApp numbers, bot
export const P_CONTACTS: Permission = "org:contacts:manage";
export const P_LEADS_VIEW_ALL: Permission = "org:leads:view_all"; // see every lead in the workspace
export const P_LEADS_ASSIGN: Permission = "org:leads:assign"; // assign/reassign leads
export const P_REPORTS: Permission = "org:reports:view";

/** All eleven permissions, in catalogue order. */
export const PERMISSIONS: readonly Permission[] = [
  P_BILLING,
  P_WORKSPACE_EDIT,
  P_WORKSPACE_DEL,
  P_TEAM_MANAGE,
  P_CAMPAIGNS,
  P_TEMPLATES,
  P_SETTINGS,
  P_CONTACTS,
  P_LEADS_VIEW_ALL,
  P_LEADS_ASSIGN,
  P_REPORTS,
] as const;

// Per-role grants. `org:owner` is handled specially (all permissions).
export const ROLE_PERMISSIONS: Record<Role, ReadonlySet<Permission>> = {
  "org:owner": new Set<Permission>(), // special-cased to "everything" in hasPermission
  "org:admin": new Set<Permission>([
    P_TEAM_MANAGE,
    P_CAMPAIGNS,
    P_TEMPLATES,
    P_SETTINGS,
    P_CONTACTS,
    P_LEADS_VIEW_ALL,
    P_LEADS_ASSIGN,
    P_REPORTS,
  ]),
  "org:manager": new Set<Permission>([
    P_LEADS_VIEW_ALL,
    P_LEADS_ASSIGN,
    P_REPORTS,
    P_CONTACTS,
  ]),
  "org:agent": new Set<Permission>(), // only their assigned leads/conversations (no view_all)
  "org:viewer": new Set<Permission>([P_REPORTS]), // read-only dashboards
};

/** True if `role` is granted `permission` (owner has all). */
export function hasPermission(
  role: Role | null | undefined,
  permission: Permission,
): boolean {
  if (role === "org:owner") {
    return true;
  }
  if (!role) {
    return false;
  }
  return ROLE_PERMISSIONS[role]?.has(permission) ?? false;
}

export function canViewAllLeads(role: Role | null | undefined): boolean {
  return hasPermission(role, P_LEADS_VIEW_ALL);
}

export function canAssignLeads(role: Role | null | undefined): boolean {
  return hasPermission(role, P_LEADS_ASSIGN);
}

/**
 * An actor may set/manage a member only at or below their own rank, and only
 * owners may grant or manage the `org:owner` role. Port of can_manage_role.
 */
export function canManageRole(
  actorRole: Role | null | undefined,
  targetRole: Role | null | undefined,
): boolean {
  if (!hasPermission(actorRole, P_TEAM_MANAGE)) {
    return false;
  }
  if (targetRole === "org:owner" && actorRole !== "org:owner") {
    return false;
  }
  const actorRank = actorRole ? ROLE_RANK[actorRole] : -1;
  const targetRank = targetRole ? ROLE_RANK[targetRole] : 99;
  return actorRank >= targetRank;
}

/** Resolve all effective permissions for a role (owner ⇒ the full catalogue). */
export function permissionsForRole(
  role: Role | null | undefined,
): ReadonlySet<Permission> {
  if (role === "org:owner") {
    return new Set<Permission>(PERMISSIONS);
  }
  if (!role) {
    return new Set<Permission>();
  }
  return ROLE_PERMISSIONS[role] ?? new Set<Permission>();
}

File: backend/auth/rbac.py
"""
Role-based access control for workspace members.

Five roles, highest → lowest privilege. Each maps to a set of permission
strings; ``owner`` implicitly has every permission. Permissions are checked at
the API boundary via ``auth.deps.require_permission``.

    owner    full control incl. billing, delete workspace, manage admins
    admin    manage team, campaigns, templates, settings
    manager  view all leads, assign leads, see reports
    agent    handle only leads/conversations assigned to them
    viewer   read-only
"""
from __future__ import annotations

# Privilege rank — used to stop a member from managing someone above them.
ROLE_RANK = {"owner": 4, "admin": 3, "manager": 2, "agent": 1, "viewer": 0}
ROLES = tuple(ROLE_RANK.keys())

# Permission catalogue (granular capabilities used by endpoints).
P_BILLING        = "billing.manage"
P_WORKSPACE_EDIT = "workspace.edit"
P_WORKSPACE_DEL  = "workspace.delete"
P_TEAM_MANAGE    = "team.manage"          # invite/remove members, change roles, sales teams
P_CAMPAIGNS      = "campaigns.manage"
P_TEMPLATES      = "templates.manage"
P_SETTINGS       = "settings.manage"      # vertical, WhatsApp numbers, bot
P_CONTACTS       = "contacts.manage"
P_LEADS_VIEW_ALL = "leads.view_all"       # see every lead in the workspace
P_LEADS_ASSIGN   = "leads.assign"         # assign/reassign leads
P_REPORTS        = "reports.view"

# Per-role grants. ``owner`` is handled specially (all permissions).
ROLE_PERMISSIONS: dict[str, set[str]] = {
    "owner": set(),  # special-cased to "everything" in has_permission
    "admin": {
        P_TEAM_MANAGE, P_CAMPAIGNS, P_TEMPLATES, P_SETTINGS, P_CONTACTS,
        P_LEADS_VIEW_ALL, P_LEADS_ASSIGN, P_REPORTS,
    },
    "manager": {
        P_LEADS_VIEW_ALL, P_LEADS_ASSIGN, P_REPORTS, P_CONTACTS,
    },
    "agent": set(),       # only their assigned leads/conversations (no view_all)
    "viewer": {P_REPORTS},  # read-only dashboards
}


def has_permission(role: str | None, permission: str) -> bool:
    """True if ``role`` is granted ``permission`` (owner has all)."""
    if role == "owner":
        return True
    return permission in ROLE_PERMISSIONS.get(role or "", set())


def can_view_all_leads(role: str | None) -> bool:
    return has_permission(role, P_LEADS_VIEW_ALL)


def can_assign_leads(role: str | None) -> bool:
    return has_permission(role, P_LEADS_ASSIGN)


def can_manage_role(actor_role: str | None, target_role: str | None) -> bool:
    """An actor may set/manage a member only at or below their own rank, and
    only owners may grant or manage the ``owner`` role."""
    if not has_permission(actor_role, P_TEAM_MANAGE):
        return False
    if target_role == "owner" and actor_role != "owner":
        return False
    return ROLE_RANK.get(actor_role or "", -1) >= ROLE_RANK.get(target_role or "", 99)
